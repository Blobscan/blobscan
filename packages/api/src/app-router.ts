import { z } from "@blobscan/zod";

import { publicProcedure } from "./procedures";
import { blobRouter } from "./routers/blob";
import { blobStoragesStateRouter } from "./routers/blob-storages-state";
import { blockRouter } from "./routers/block";
import { blockchainSyncStateRouter } from "./routers/blockchain-sync-state";
import { indexerRouter } from "./routers/indexer";
import { searchRouter } from "./routers/search";
import { statsRouter } from "./routers/stats";
import { transactionRouter } from "./routers/tx";
import { t } from "./trpc-client";

export const appRouter = t.router({
  block: blockRouter,
  tx: transactionRouter,
  blob: blobRouter,
  search: searchRouter,
  stats: statsRouter,
  indexer: indexerRouter,
  syncState: blockchainSyncStateRouter,
  blobStoragesState: blobStoragesStateRouter,
  healthcheck: publicProcedure
    .meta({
      openapi: {
        method: "GET",
        path: "/healthcheck",
        summary: "connection healthcheck.",
        tags: ["system"],
      },
    })
    .input(z.void())
    .output(z.string())
    .query(() => "yay!"),
});

// export type definition of API
export type AppRouter = typeof appRouter;
