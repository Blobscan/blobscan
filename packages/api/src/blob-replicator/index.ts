import { FlowProducer, Worker } from "bullmq";
import type { FlowChildJob, FlowJob, WorkerOptions } from "bullmq";
import path from "path";

import { createOrLoadBlobStorageManager } from "@blobscan/blob-storage-manager";
import { BlobStorage } from "@blobscan/db";

import { FLOW_PRODUCER_QUEUE, STORAGE_QUEUES } from "./config";
import type { BlobReplicationJobData } from "./types";

const WORKERS_DIR = "workers";

const STORAGE_WORKER_FILES: Record<BlobStorage, string> = {
  GOOGLE: path.join(__dirname, WORKERS_DIR, "gcs.js"),
  SWARM: path.join(__dirname, WORKERS_DIR, "swarm.js"),
  POSTGRES: path.join(__dirname, WORKERS_DIR, "postgres.js"),
};
const PROPAGATOR_WORKER_FILE = path.join(
  __dirname,
  WORKERS_DIR,
  "propagator-flow-producer.js"
);

const replicatorFlowProducer = new FlowProducer();

const BLOB_STORAGES = Object.values(BlobStorage);

export async function setUpBlobReplicationWorkers() {
  const blobStorageManager = await createOrLoadBlobStorageManager();
  const workerOptions: WorkerOptions = {
    useWorkerThreads: true,
  };

  const storageWorkers = BLOB_STORAGES.filter(
    blobStorageManager.hasStorage
  ).reduce<Record<BlobStorage, Worker>>((workers, storageName) => {
    const storageWorker = new Worker(
      STORAGE_QUEUES[storageName],
      STORAGE_WORKER_FILES[storageName],
      workerOptions
    );

    storageWorker.on("completed", (job) => {
      console.log(`Completed job ${job.id} in ${STORAGE_QUEUES[storageName]}`);
    });

    storageWorker.on("failed", (job, err) => {
      console.error(
        `Failed job ${job?.id} in ${STORAGE_QUEUES[storageName]}: ${err}`
      );
    });

    workers[storageName] = storageWorker;

    return workers;
  }, {} as Record<BlobStorage, Worker>);

  if (Object.values(storageWorkers).length) {
    const replicatorWorker = new Worker(
      FLOW_PRODUCER_QUEUE,
      PROPAGATOR_WORKER_FILE,
      workerOptions
    );
  }
}

async function createBlobReplicationJob(
  data: BlobReplicationJobData
): Promise<FlowJob> {
  const blobStorageManager = await createOrLoadBlobStorageManager();
  const versionedHash = data.versionedHash;

  const storageJobs: FlowChildJob[] = BLOB_STORAGES.filter(
    blobStorageManager.hasStorage
  ).map<FlowChildJob>((storage) => {
    const jobId = `${storage}-${versionedHash}`;

    return {
      name: jobId,
      queueName: STORAGE_QUEUES[storage],
      data,
      opts: {
        jobId,
      },
    };
  });

  const jobId = `${FLOW_PRODUCER_QUEUE}-${versionedHash}`;

  return {
    name: jobId,
    queueName: FLOW_PRODUCER_QUEUE,
    data,
    opts: {
      jobId,
    },
    children: storageJobs,
  };
}

export async function queueBlobForReplication(data: BlobReplicationJobData) {
  const blobFlowJob = await createBlobReplicationJob(data);

  replicatorFlowProducer.add(blobFlowJob);
}

export async function queueBlobsForReplication(
  blobReplicationData: BlobReplicationJobData[]
) {
  const blobReplicationFlowJobs = await Promise.all(
    blobReplicationData.map((data) => createBlobReplicationJob(data))
  );

  return replicatorFlowProducer.addBulk(blobReplicationFlowJobs);
}
