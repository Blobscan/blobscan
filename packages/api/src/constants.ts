export const DEFAULT_LIMIT = 50;
