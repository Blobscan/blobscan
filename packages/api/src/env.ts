import { env as blobPropagatorEnv } from "@blobscan/blob-propagator";
import { env as blobStorageManagerEnv } from "@blobscan/blob-storage-manager";
import { env as openTelemetryEnv } from "@blobscan/open-telemetry";
import {
  z,
  createEnv,
  presetEnvOptions,
  nodeEnvSchema,
  booleanSchema,
  maskSensitiveData,
} from "@blobscan/zod";

export const env = createEnv({
  envOptions: {
    server: {
      CHAIN_ID: z.coerce.number().positive().default(1),
      SECRET_KEY: z.string(),
      NODE_ENV: nodeEnvSchema.optional(),
      METRICS_ENABLED: booleanSchema.default("false"),
    },

    ...presetEnvOptions,
  },
  display(env) {
    console.log(
      `API configuration: secretKey: ${maskSensitiveData(env.SECRET_KEY)}`
    );
    blobPropagatorEnv.display();
    blobStorageManagerEnv.display();
    openTelemetryEnv.display();
  },
});
