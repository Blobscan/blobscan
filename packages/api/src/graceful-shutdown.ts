import { blobPropagator } from "@blobscan/blob-propagator";
import { prisma } from "@blobscan/db";

export async function gracefulShutdown() {
  const shutdownOps = [];

  if (blobPropagator) {
    shutdownOps.push(blobPropagator.close());
  }

  shutdownOps.push(prisma.$disconnect());

  return Promise.all(shutdownOps);
}
