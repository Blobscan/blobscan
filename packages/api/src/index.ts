import type { inferRouterInputs, inferRouterOutputs } from "@trpc/server";

import type { AppRouter } from "./app-router";

export { appRouter, type AppRouter } from "./app-router";
export { createTRPCContext, createTRPCInnerContext } from "./context";
export type {
  TRPCInnerContext as TRCInnerContext,
  TRPCContext,
} from "./context";

export * from "@trpc/server/adapters/express";

export { env } from "./env";
export { gracefulShutdown } from "./graceful-shutdown";
export { metricsHandler } from "./instrumentation";
export type {
  BlobStorage,
  Rollup,
  Metric,
  MetricHistogram,
  MetricHistogramBucket,
  Metrics,
  MetricsClient,
} from "@blobscan/db";

/**
 * Inference helpers for input types
 * @example type HelloInput = RouterInputs['example']['hello']
 **/
export type RouterInputs = inferRouterInputs<AppRouter>;

/**
 * Inference helpers for output types
 * @example type HelloOutput = RouterOutputs['example']['hello']
 **/
export type RouterOutputs = inferRouterOutputs<AppRouter>;
