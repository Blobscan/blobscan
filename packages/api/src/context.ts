import { TRPCError } from "@trpc/server";
import type { inferAsyncReturnType } from "@trpc/server";
import type { CreateNextContextOptions } from "@trpc/server/adapters/next";
import type {
  NodeHTTPCreateContextFnOptions,
  NodeHTTPRequest,
  NodeHTTPResponse,
} from "@trpc/server/adapters/node-http";
import jwt from "jsonwebtoken";

import { blobPropagator } from "@blobscan/blob-propagator";
import { getBlobStorageManager } from "@blobscan/blob-storage-manager";
import { prisma } from "@blobscan/db";

import { env } from "./env";

export type CreateContextOptions =
  | NodeHTTPCreateContextFnOptions<NodeHTTPRequest, NodeHTTPResponse>
  | CreateNextContextOptions;

type CreateInnerContextOptions = Partial<CreateContextOptions> & {
  apiClient: string | null;
};

function getJWTFromRequest(
  req: NodeHTTPRequest | CreateNextContextOptions["req"]
) {
  const authHeader = req.headers.authorization;
  if (!authHeader) {
    return null;
  }

  try {
    const [type, token] = authHeader.split(" ");
    if (type !== "Bearer" || !token) {
      return null;
    }

    const decoded = jwt.verify(token, env.SECRET_KEY) as string;

    return decoded;
  } catch (err) {
    if (err instanceof jwt.JsonWebTokenError) {
      return null;
    }

    throw new TRPCError({ code: "BAD_REQUEST" });
  }
}

export async function createTRPCInnerContext(opts?: CreateInnerContextOptions) {
  const blobStorageManager = await getBlobStorageManager();

  return {
    prisma,
    blobStorageManager,
    blobPropagator,
    apiClient: opts?.apiClient,
  };
}

export function createTRPCContext(
  {
    scope,
  }: {
    scope: string;
  } = { scope: "" }
) {
  return async (opts: CreateContextOptions) => {
    try {
      const apiClient = getJWTFromRequest(opts.req);

      const innerContext = await createTRPCInnerContext({
        apiClient,
      });

      return {
        ...innerContext,
        scope,
        req: opts.req,
        res: opts.res,
      };
    } catch (err) {
      const err_ = err as Error;

      throw new TRPCError({
        code: "INTERNAL_SERVER_ERROR",
        message: `Failed to create TRPC context: ${err_.message}`,
      });
    }
  };
}

export type TRPCContext = inferAsyncReturnType<
  ReturnType<typeof createTRPCContext>
>;
export type TRPCInnerContext = inferAsyncReturnType<
  typeof createTRPCInnerContext
>;
