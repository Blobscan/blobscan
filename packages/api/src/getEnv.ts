import { env } from "@blobscan/env";

import { publicProcedure } from "./procedures";

export const getEnv = publicProcedure.query(() => {
  const clientEnv = Object.entries(env)
    .filter(([key]) => key.startsWith("NEXT_"))
    .map(([key, value]) => [key.replace(/^NEXT_/, ""), value]);

  return Object.fromEntries(clientEnv);
});
