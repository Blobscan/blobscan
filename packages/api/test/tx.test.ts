import { beforeAll, describe, expect, it } from "vitest";

import { fixtures } from "@blobscan/test";

import type { TRPCContext } from "../src";
import { appRouter } from "../src/app-router";
import {
  createTestContext,
  runExpandsTestsSuite,
  runFiltersTestsSuite,
  runPaginationTestsSuite,
} from "./helpers";

describe("Transaction router", async () => {
  let caller: ReturnType<typeof appRouter.createCaller>;
  let ctx: TRPCContext;

  beforeAll(async () => {
    ctx = await createTestContext();
    caller = appRouter.createCaller(ctx);
  });

  describe("getAll", () => {
    runPaginationTestsSuite("transaction", (paginationInput) =>
      caller.tx.getAll(paginationInput).then(({ transactions }) => transactions)
    );

    runFiltersTestsSuite("transaction", (filterInput) =>
      caller.tx.getAll(filterInput).then(({ transactions }) => transactions)
    );

    runExpandsTestsSuite("transaction", ["block", "blob"], (input) =>
      caller.tx.getAll(input).then(({ transactions }) => transactions)
    );

    it("should get the total number of transactions", async () => {
      const expectedTotalTransactions = fixtures.canonicalTxs.length;

      await ctx.prisma.transactionOverallStats.increment({
        from: 0,
        to: 9999,
      });

      const { totalTransactions } = await caller.tx.getAll();

      expect(totalTransactions).toBe(expectedTotalTransactions);
    });

    it("should get the total number of transactions for a rollup", async () => {
      const expectedTotalTransactions = await ctx.prisma.transaction.count({
        where: {
          rollup: "BASE",
        },
      });

      const { totalTransactions } = await caller.tx.getAll({
        rollup: "base",
      });

      expect(totalTransactions).toBe(expectedTotalTransactions);
    });
  });

  describe("getByHash", () => {
    runExpandsTestsSuite(
      "transaction",
      ["block", "blob", "blob_data"],
      (expandsInput) =>
        caller.tx.getByHash({ hash: "txHash001", ...expandsInput })
    );

    it("should get a transaction by hash correctly", async () => {
      const result = await caller.tx.getByHash({
        hash: "txHash001",
      });
      expect(result).toMatchSnapshot();
    });

    it("should fail when providing a non-existent hash", async () => {
      await expect(
        caller.tx.getByHash({
          hash: "nonExistingHash",
        })
      ).rejects.toMatchSnapshot(
        "[TRPCError: No transaction with hash 'nonExistingHash' found]"
      );
    });
  });
});
