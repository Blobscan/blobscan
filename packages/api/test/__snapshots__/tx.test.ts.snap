// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Transaction router > 'getAll' > when getting paginated transaction results > should default to the first page when no page was specified 1`] = `
[
  {
    "blobAsCalldataGasUsed": "1000",
    "blobGasBaseFee": "2883584",
    "blobGasMaxFee": "13107200",
    "blobGasPrice": "22",
    "blobGasUsed": "131072",
    "blobs": [
      "blobHash001",
    ],
    "blockHash": "blockHash008",
    "blockNumber": 1008,
    "from": "address2",
    "gasPrice": "10",
    "hash": "txHash016",
    "maxFeePerBlobGas": "100",
    "to": "address4",
    "totalBlobSize": 1000,
  },
  {
    "blobAsCalldataGasUsed": "1000",
    "blobGasBaseFee": "8650752",
    "blobGasMaxFee": "39321600",
    "blobGasPrice": "22",
    "blobGasUsed": "393216",
    "blobs": [
      "blobHash004",
      "blobHash004",
      "blobHash004",
    ],
    "blockHash": "blockHash007",
    "blockNumber": 1007,
    "from": "address5",
    "gasPrice": "10",
    "hash": "txHash015",
    "maxFeePerBlobGas": "100",
    "to": "address4",
    "totalBlobSize": 3900,
  },
]
`;

exports[`Transaction router > 'getAll' > when getting paginated transaction results > should return the correct number of results 1`] = `
[
  {
    "blobAsCalldataGasUsed": "1000",
    "blobGasBaseFee": "2883584",
    "blobGasMaxFee": "13107200",
    "blobGasPrice": "22",
    "blobGasUsed": "131072",
    "blobs": [
      "blobHash001",
    ],
    "blockHash": "blockHash008",
    "blockNumber": 1008,
    "from": "address2",
    "gasPrice": "10",
    "hash": "txHash016",
    "maxFeePerBlobGas": "100",
    "to": "address4",
    "totalBlobSize": 1000,
  },
  {
    "blobAsCalldataGasUsed": "1000",
    "blobGasBaseFee": "8650752",
    "blobGasMaxFee": "39321600",
    "blobGasPrice": "22",
    "blobGasUsed": "393216",
    "blobs": [
      "blobHash004",
      "blobHash004",
      "blobHash004",
    ],
    "blockHash": "blockHash007",
    "blockNumber": 1007,
    "from": "address5",
    "gasPrice": "10",
    "hash": "txHash015",
    "maxFeePerBlobGas": "100",
    "to": "address4",
    "totalBlobSize": 3900,
  },
]
`;

exports[`Transaction router > 'getAllFull' > when getting paginated transaction results > should default to the first page when no page was specified 1`] = `
[
  {
    "blobAsCalldataGasUsed": "1000",
    "blobGasBaseFee": "2883584",
    "blobGasMaxFee": "13107200",
    "blobGasUsed": "131072",
    "blobs": [
      {
        "blob": {
          "commitment": "commitment001",
          "proof": "proof001",
          "size": 1000,
        },
        "blobHash": "blobHash001",
        "index": 0,
      },
    ],
    "block": {
      "blobGasPrice": "22",
      "excessBlobGas": "15000",
      "number": 1008,
      "timestamp": 1693497600,
    },
    "blockHash": "blockHash008",
    "fromId": "address2",
    "gasPrice": "10",
    "hash": "txHash016",
    "maxFeePerBlobGas": "100",
    "toId": "address4",
    "totalBlobSize": 1000,
  },
  {
    "blobAsCalldataGasUsed": "1000",
    "blobGasBaseFee": "8650752",
    "blobGasMaxFee": "39321600",
    "blobGasUsed": "393216",
    "blobs": [
      {
        "blob": {
          "commitment": "commitment004",
          "proof": "proof004",
          "size": 1300,
        },
        "blobHash": "blobHash004",
        "index": 0,
      },
      {
        "blob": {
          "commitment": "commitment004",
          "proof": "proof004",
          "size": 1300,
        },
        "blobHash": "blobHash004",
        "index": 1,
      },
      {
        "blob": {
          "commitment": "commitment004",
          "proof": "proof004",
          "size": 1300,
        },
        "blobHash": "blobHash004",
        "index": 2,
      },
    ],
    "block": {
      "blobGasPrice": "22",
      "excessBlobGas": "15000",
      "number": 1007,
      "timestamp": 1693490400,
    },
    "blockHash": "blockHash007",
    "fromId": "address5",
    "gasPrice": "10",
    "hash": "txHash015",
    "maxFeePerBlobGas": "100",
    "toId": "address4",
    "totalBlobSize": 3900,
  },
]
`;

exports[`Transaction router > 'getAllFull' > when getting paginated transaction results > should return the correct number of results 1`] = `
[
  {
    "blobAsCalldataGasUsed": "1000",
    "blobGasBaseFee": "2883584",
    "blobGasMaxFee": "13107200",
    "blobGasUsed": "131072",
    "blobs": [
      {
        "blob": {
          "commitment": "commitment001",
          "proof": "proof001",
          "size": 1000,
        },
        "blobHash": "blobHash001",
        "index": 0,
      },
    ],
    "block": {
      "blobGasPrice": "22",
      "excessBlobGas": "15000",
      "number": 1008,
      "timestamp": 1693497600,
    },
    "blockHash": "blockHash008",
    "fromId": "address2",
    "gasPrice": "10",
    "hash": "txHash016",
    "maxFeePerBlobGas": "100",
    "toId": "address4",
    "totalBlobSize": 1000,
  },
  {
    "blobAsCalldataGasUsed": "1000",
    "blobGasBaseFee": "8650752",
    "blobGasMaxFee": "39321600",
    "blobGasUsed": "393216",
    "blobs": [
      {
        "blob": {
          "commitment": "commitment004",
          "proof": "proof004",
          "size": 1300,
        },
        "blobHash": "blobHash004",
        "index": 0,
      },
      {
        "blob": {
          "commitment": "commitment004",
          "proof": "proof004",
          "size": 1300,
        },
        "blobHash": "blobHash004",
        "index": 1,
      },
      {
        "blob": {
          "commitment": "commitment004",
          "proof": "proof004",
          "size": 1300,
        },
        "blobHash": "blobHash004",
        "index": 2,
      },
    ],
    "block": {
      "blobGasPrice": "22",
      "excessBlobGas": "15000",
      "number": 1007,
      "timestamp": 1693490400,
    },
    "blockHash": "blockHash007",
    "fromId": "address5",
    "gasPrice": "10",
    "hash": "txHash015",
    "maxFeePerBlobGas": "100",
    "toId": "address4",
    "totalBlobSize": 3900,
  },
]
`;

exports[`Transaction router > 'getByHash' > should fail when providing a non-existent hash > [TRPCError: No transaction with hash 'nonExistingHash'.] 1`] = `[TRPCError: No transaction with hash 'nonExistingHash'.]`;

exports[`Transaction router > 'getByHash' > should get a transaction by hash correctly 1`] = `
{
  "blobAsCalldataGasUsed": "1000",
  "blobGasBaseFee": "8650752",
  "blobGasMaxFee": "39321600",
  "blobGasPrice": "22",
  "blobGasUsed": "393216",
  "blobs": [
    "blobHash001",
    "blobHash002",
    "blobHash003",
  ],
  "blockHash": "blockHash001",
  "blockNumber": 1001,
  "from": "address1",
  "gasPrice": "10",
  "hash": "txHash001",
  "maxFeePerBlobGas": "100",
  "to": "address2",
  "totalBlobSize": 3300,
}
`;

exports[`Transaction router > 'getByHashFull' > should fail when providing a non-existent hash > [TRPCError: No transaction with hash 'nonExistingHash'.] 1`] = `[TRPCError: No transaction with hash 'nonExistingHash'.]`;

exports[`Transaction router > 'getByHashFull' > should get a transaction by hash correctly 1`] = `
{
  "blobAsCalldataGasUsed": "1000",
  "blobGasBaseFee": "8650752",
  "blobGasMaxFee": "39321600",
  "blobGasUsed": "393216",
  "blobs": [
    {
      "blob": {
        "commitment": "commitment001",
        "proof": "proof001",
        "size": 1000,
      },
      "blobHash": "blobHash001",
      "index": 0,
    },
    {
      "blob": {
        "commitment": "commitment002",
        "proof": "proof002",
        "size": 1100,
      },
      "blobHash": "blobHash002",
      "index": 1,
    },
    {
      "blob": {
        "commitment": "commitment003",
        "proof": "proof003",
        "size": 1200,
      },
      "blobHash": "blobHash003",
      "index": 2,
    },
  ],
  "block": {
    "blobGasPrice": "22",
    "excessBlobGas": "15000",
    "number": 1001,
    "timestamp": 1665921600,
  },
  "blockHash": "blockHash001",
  "fromId": "address1",
  "gasPrice": "10",
  "hash": "txHash001",
  "maxFeePerBlobGas": "100",
  "toId": "address2",
  "totalBlobSize": 3300,
}
`;

exports[`Transaction router > getByAddress > when getting paginated address's transactions results > should default to the first page when no page was specified 1`] = `
[
  {
    "blobAsCalldataGasUsed": "1000",
    "blobGasBaseFee": "2883584",
    "blobGasMaxFee": "13107200",
    "blobGasUsed": "131072",
    "blobs": [
      {
        "blob": {
          "commitment": "commitment001",
          "proof": "proof001",
          "size": 1000,
        },
        "blobHash": "blobHash001",
        "index": 0,
      },
    ],
    "block": {
      "blobGasPrice": "22",
      "excessBlobGas": "15000",
      "number": 1008,
      "timestamp": 1693497600,
    },
    "blockHash": "blockHash008",
    "fromId": "address2",
    "gasPrice": "10",
    "hash": "txHash016",
    "maxFeePerBlobGas": "100",
    "toId": "address4",
    "totalBlobSize": 1000,
  },
  {
    "blobAsCalldataGasUsed": "1000",
    "blobGasBaseFee": "7864320",
    "blobGasMaxFee": "39321600",
    "blobGasUsed": "393216",
    "blobs": [
      {
        "blob": {
          "commitment": "commitment001",
          "proof": "proof001",
          "size": 1000,
        },
        "blobHash": "blobHash001",
        "index": 0,
      },
      {
        "blob": {
          "commitment": "commitment001",
          "proof": "proof001",
          "size": 1000,
        },
        "blobHash": "blobHash001",
        "index": 1,
      },
      {
        "blob": {
          "commitment": "commitment001",
          "proof": "proof001",
          "size": 1000,
        },
        "blobHash": "blobHash001",
        "index": 2,
      },
    ],
    "block": {
      "blobGasPrice": "20",
      "excessBlobGas": "10000",
      "number": 1005,
      "timestamp": 1693216800,
    },
    "blockHash": "blockHash005",
    "fromId": "address2",
    "gasPrice": "10",
    "hash": "txHash012",
    "maxFeePerBlobGas": "100",
    "toId": "address4",
    "totalBlobSize": 3000,
  },
]
`;

exports[`Transaction router > getByAddress > when getting paginated address's transactions results > should return the correct number of results 1`] = `
[
  {
    "blobAsCalldataGasUsed": "1000",
    "blobGasBaseFee": "2883584",
    "blobGasMaxFee": "13107200",
    "blobGasUsed": "131072",
    "blobs": [
      {
        "blob": {
          "commitment": "commitment001",
          "proof": "proof001",
          "size": 1000,
        },
        "blobHash": "blobHash001",
        "index": 0,
      },
    ],
    "block": {
      "blobGasPrice": "22",
      "excessBlobGas": "15000",
      "number": 1008,
      "timestamp": 1693497600,
    },
    "blockHash": "blockHash008",
    "fromId": "address2",
    "gasPrice": "10",
    "hash": "txHash016",
    "maxFeePerBlobGas": "100",
    "toId": "address4",
    "totalBlobSize": 1000,
  },
  {
    "blobAsCalldataGasUsed": "1000",
    "blobGasBaseFee": "7864320",
    "blobGasMaxFee": "39321600",
    "blobGasUsed": "393216",
    "blobs": [
      {
        "blob": {
          "commitment": "commitment001",
          "proof": "proof001",
          "size": 1000,
        },
        "blobHash": "blobHash001",
        "index": 0,
      },
      {
        "blob": {
          "commitment": "commitment001",
          "proof": "proof001",
          "size": 1000,
        },
        "blobHash": "blobHash001",
        "index": 1,
      },
      {
        "blob": {
          "commitment": "commitment001",
          "proof": "proof001",
          "size": 1000,
        },
        "blobHash": "blobHash001",
        "index": 2,
      },
    ],
    "block": {
      "blobGasPrice": "20",
      "excessBlobGas": "10000",
      "number": 1005,
      "timestamp": 1693216800,
    },
    "blockHash": "blockHash005",
    "fromId": "address2",
    "gasPrice": "10",
    "hash": "txHash012",
    "maxFeePerBlobGas": "100",
    "toId": "address4",
    "totalBlobSize": 3000,
  },
]
`;
